Add begin() method

dok-net opened the pull request:
It's usual for Arduino driver libs to have a constructor suitable for storage class static, and a `begin()` method that's called from `setup()`. The I2C address can thus be set in either way.
This PR adds that function.

dok-net committed:
Add begin() method usual for Arduino driver libs and the supporting ctor.

blemasle commented:
Hi, and thanks for the contribution 🙂 

I'm not sure about this one though. As stated in the [README](https://github.com/blemasle/arduino-mcp23017#usage)
> Unlike most Arduino library, no default instance is created when the library is included

The usual `begin` method is here to compensate for the lack of a proper constructor because libraries usually declare a unique, static instance.

I agree that the I2C address can be set either way with your PR, but it also means that you can create an instance and start using it without ever setting the address, which would be error prone. It also means that you can set this address using the constructor, and then use `begin()` without any argument which will overwrite that value.

Moreover, most of the libs I've seen use `begin()` to actually _begin_ the communication with the device/port. Eg `Serial.begin` actually opens the serial port. Here, nothing would "begin" when you call that function.

What do you think ? 🙂

dok-net committed:
Begin without address argument shall use the ctor provided address, not reset it.

dok-net commented:
Hi. Obviously, I have a different take :-)
First, the current pattern isn't obsoleted by my patch.
Second, there are libraries don't create a global default instance and still offer `begin()`.
Third, I don't subscribe to the danger argument, as when using the ctor without a specified address value,
the default address gets set, that is, the base I2C address when the address lines are pulled down, which I would
like to believed is used a lot in applications. It's only after `begin()` or `init()` that there is an actual attempt to access the device by that address.

As for
> It also means that you can set this address using the constructor, and then use begin() without any argument which will overwrite that value.

You're right, I missed that, I have fixed that now.

Is this PR better suited for inclusion now?

blemasle commented on src/MCP23017.cpp:
```diff
@@ -25,6 +30,17 @@ void MCP23017::init()
+
+void MCP23017::begin(uint8_t address)
+{
+	_deviceAddr = address;
+	init();
```
```suggestion
	_deviceAddr = address;
	begin();
```

In case `begin` would change in the future, it's better to call the other signature than duplicating code. 🙂

blemasle commented on src/MCP23017.h:
```diff
@@ -65,10 +66,23 @@ class MCP23017
+	 */
+	void begin();
+	/**
+	 * Overrides the I2C address set by the constructor. Implicitly calls init().
```
```suggestion
	 * Overrides the I2C address set by the constructor. Implicitly calls begin().

```

blemasle reviewed (commented):
Alright, you convinced me 🙂 Despite the fact that I still think the `begin()` pattern used by a lot of libraries is misleading and confusing, I get the "everything else works that way" argument.

Would you review the suggestions I made though ?

dok-net committed:
Apply suggestions from code review

Cascade from `begin` with I2C address argument to `begin` w/o argument, it's clearer than calling `init()` from both places.

Co-authored-by: Bertrand Lemasle <blemasle@gmail.com>

dok-net commented:
Thanks for the nice review feedback.
